src/bluez/adapter/peripheral.rs

```diff
@@ -11,410 +11,404 @@
-                let mut stream = self.stream.write().unwrap();
-                *stream = None;
-                // TODO clean up our sockets
+            cvar.notify_all();
+        }
+
+        if let Some(name) = args.get("Name") {
+            let name = name.as_str().map(|s| s.to_string());
+            debug!("Updating \"{}\" local name to \"{:?}\"", self.address, name);
+            properties.local_name = name;
+        }
+
+        if let Some(services_resolved) = args.get("ServicesResolved") {
+            let services_resolved = services_resolved.0.as_u64().unwrap() > 0;
+            if services_resolved {
+                // Need to prase and figure out handle ranges for all discovered characteristics.
+                self.build_characteristic_ranges().unwrap();
             }
-            msg => {
-                debug!("ignored message {:?}", msg);
+            // All services have been discovered, time to inform anyone waiting.
+            let (ref lock, ref cvar) = *self.state;
+            if services_resolved {
+                *lock.lock().unwrap() = PeripheralState::ServicesResolved;
             }
+            cvar.notify_all();
         }
-    }
 
-    fn request_raw_async(&self, data: &mut [u8], handler: Option<RequestCallback>) {
-        let l = self.stream.read().unwrap();
-        match l.as_ref().ok_or(Error::NotConnected) {
-            Ok(stream) => {
-                stream.write(&mut *data, handler);
-            }
-            Err(err) => {
-                if let Some(h) = handler {
-                    h(Err(err));
+        // if let Some(services) = args.get("ServiceData") {
+        //     debug!("Updating services to \"{:?}\"", services);
+
+        //     if let Some(mut iter) = services.0.as_iter() {
+        //         loop {
+        //             if let (Some(uuid), ())
+        //         }
+        //     }
+        // }
+
+        // As of writing this: ManufacturerData returns a 'Variant({<manufacturer_id>: Variant([<manufacturer_data>])})'.
+        // This Variant wrapped dictionary and array is difficult to navigate. So uh.. trust me, this works on my machine™.
+        if let Some(manufacturer_data) = args.get("ManufacturerData") {
+            debug!(
+                "Updating \"{}\" manufacturer data \"{:?}\"",
+                self.address, manufacturer_data
+            );
+            let mut result = Vec::<u8>::new();
+            // dbus-rs doesn't really have a dictionary API... so need to iterate two at a time and make a key-value pair.
+            if let Some(mut iter) = manufacturer_data.0.as_iter() {
+                loop {
+                    if let (Some(id), Some(data)) = (iter.next(), iter.next()) {
+                        // This API is terrible.. why can't I just get an array out, why is it all wrapped in a Variant?
+                        let data: Vec<u8> = data
+                            .as_iter() // 🎶 The Variant is connected to the
```

Comment:
I like this, please go through the rest of your patch and make your comments rhyme.